### PR TITLE
fix the grad_acc issue at epoch boundaries

### PR DESCRIPTION
**Pull request by pacman100**
# What does this PR do?

Should solve accumulating via epoch when using Accelerate (seen in https://github.com/huggingface/transformers/issues/23935#issuecomment-1588134562). Requires https://github.com/huggingface/accelerate/pull/1624

Fixes # (issue)

### CODE REVIEW

**Comment by nivibilla** [comment]
I believe this PR also fixes #24245

**Review by muellerzr — APPROVED** [review]
Thanks!

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing! 

Happy with the changes. My only concern is things breaking as this relies on feature that's just been merged on `accelerate`'s main branch. Is the next release of accelerate going to be before the next release of transformers? 

@pacman100 @muellerzr You've been handling the switch using accelerate in trainer really well, and I can see our CI runs install accelerate from source, so happy if this follows the pattern you've using.

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1812,14 +1812,21 @@ def _inner_training_loop(
...
+                    or
                     # last step in epoch but step is always smaller than gradient_accumulation_steps
-                    steps_in_epoch <= args.gradient_accumulation_steps
-                    and (step + 1) == steps_in_epoch
+                    is_last_step_and_steps_less_than_grad_acc
                 ):
+                    # the `or` condition of `is_last_step_and_steps_less_than_grad_acc` is not covered
+                    # in accelerate. So, explicitly enable sync gradients to True in that case.
+                    if is_last_step_and_steps_less_than_grad_acc:
+                        self.accelerator.gradient_state._set_sync_gradients(True)
```
Comment (sgugger) [inline src/transformers/trainer.py:1828]: I think this API is only available on main in Accelerate, so there should be a check here until next release of Accelerate.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -3842,10 +3849,13 @@ def _add_sm_patterns_to_gitignore(self) -> None:
             self.repo.git_push()
 
     def create_accelerator_and_postprocess(self):
+        gradient_accumulation_plugin = GradientAccumulationPlugin(
+            num_steps=self.args.gradient_accumulation_steps, sync_with_dataloader=False
```
Comment (sgugger) [inline src/transformers/trainer.py:3853]: Same here for `sync_with_dataloader`

**Review by sgugger — APPROVED** [review]
Thanks for the work on this! There needs to be a couple of protections before this is merged (that we can remove at the next release of Accelerate when we move the pinned version up).

**Comment by Oxi84** [comment]
Hello,

How do i install this? I expected that PR means some kind of transformers update, in this case there should be install link such as git+https://github.com/huggingface/transformers@de9255de27abfcae4a1f816b904915f0b1e23cd9

File: src/transformers/trainer.py (hunk 1 above)
Comment (pacman100) [reply src/transformers/trainer.py:1828]: this is available in the release version

File: src/transformers/trainer.py (hunk 2 above)
Comment (pacman100) [reply src/transformers/trainer.py:3853]: this is available on main

Comment (pacman100) [reply src/transformers/trainer.py]: done

File: src/transformers/trainer.py (hunk 1 above)
Comment (pacman100) [reply src/transformers/trainer.py:1830]: done

**Comment by pacman100** [comment]
Hello @Oxi84, you can install this once it gets merged via `pip install git+https://github.com/huggingface/transformers` and `pip install git+https://github.com/huggingface/accelerate`

**Review by sgugger — APPROVED** [review]

**Comment by philpax** [comment]
Just completed a training run with this PR and can confirm that the issue didn't occur. Thanks for the fix!

**Pull request merged**